Do not rely on ConsulApi retries in consul namer

edio opened the pull request:
Consul namer uses ConsulApi with retry=true to
retry on certain types of errors, backoffs is used
by ConsulApi while retrying. However if error is
not handled by ConsulApi, it is propagated as
exception to SvcAddr and retry is attemted there
without backoff. This is potentially dangerous.

This commit delegates retires fully to SvcAddr

Fixes #1821

Signed-off-by: Dmytro Kostiuchenko <kostiuchenkod2@ncbi.nlm.nih.gov>

edio commented:
All errors are handled in the same fashion: if previous state exists, we treat error as intermittent and continue watching the service. If no previous state exists, we terminate immediately.

Main differences in behavior
1) On connection exception we no longer try observing with the same blocking index. Assumption is that while consul was unreachable, service catalog state may have changed anyway, so it is safe to retry with index. And the cost is just one single request

2) On request to non-existent DC we no longer monitor address infinitely. We do this only if succeeded in the past.

3) All errors are retried with back-off

4) Only first error in a row for each name is logged as WARNING, rest are logged as DEBUG to reduce noise in logs.

adleong commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -61,7 +65,7 @@ private[consul] object SvcAddr {
       @volatile var lastGood: Option[Addr] = None
       @volatile var stopped: Boolean = false
-      def loop(index0: Option[String]): Future[Unit] = {
+      def loop(index0: Option[String], backoffs: Option[Stream[Duration]], nonFatalLogLevel: Level = Level.WARNING): Future[Unit] = {
```
Can we make this always take a `Stream[Duration]` instead of an `Option[Stream[Duration]]` and pick the default backoffs at the call sites where appropriate?  Similarly, I think we should not provide a default log level here and be more explicit about the log level at the call sites.   Since this is a local method, this should be pretty easy.

adleong commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -76,29 +80,23 @@ private[consul] object SvcAddr {
                   datacenter, key.name, e
                 )
-                loop(index0)
+                sleepAndDo(backoffs.getOrElse(apiBackoffs)) { nextBackoffs =>
```
I think it's more idiomatic to do something like

```
val backoff #:: nextBackoffs = backoffs
Future.sleep(backoff).before(loop(None, nextBackoffs, Level.DEBUG))
```

adleong commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -76,29 +80,23 @@ private[consul] object SvcAddr {
+                  // retry and log subsequent errors as debug until state changes
+                  loop(None, Some(nextBackoffs), Level.DEBUG)
+                }
               case None =>
```
I'm not sure that is makes sense to die here if there is no previous good state, especially given that we are removing the retry filter.  In the case of an error when there is no previous good state, I think we should not set `state()` (e.g. leave it as pending) and do a retry with backoff, just like the case above.  

@hawkw what do you think?

edio replied:
I was having a specific case in mind: we have security scanner running periodically (I recall seeing someone else among linkerd users with similar setup). Every now and then scanner queries for some random addresses. I'd prefere those to not be accumulated over time causing unnecessary load.
Also requests to nonexistent DCs, I believe, shouldn't be retried (although, they were previously --- in infinite retry filter). It's seems unlikely, that there are envs where DC in consul may suddenly appear after request came and thus transitioning to pending with retries will likely give no benefit ever.

adleong replied:
Fair points.  It's hard to differentiate between an invalid request (such as from a security scanner or for a nonexistent DC) from a transient failure that should be retried.  Do you know what Linkerd's behavior is when it falls into this case?  Will it cache the Addr.Failed and continue to fail subsequent requests without retrying the consul API call?  If so, I don't think that's the behavior that we want since it would effectively make transient failures put Linkerd into a "stuck" state for that name.

Regarding unnecessary load, Linkerd will stop watching these names once they have been idle for 10 minutes, so the load won't be unbounded.  And hopefully the exponential backoff will mean that the added load from these API calls will be not too much.

edio replied:
@adleong ugh, I didn't think of that. No I do not know what is the linkerd behavior.
I'll check this later.

edio replied:
I looked, how Addr.Failed will be handled by Linkerd. Here what happens to `Addr.Failed` in case if mesh interpreter is used:

1. `SvcAddr`: returns `Addr.Failed`
2. `LookupCache`: transforms it into `Activity.Failed[NameTree[Name]]`. Value gets cached through Memoize (sic!)
3. `ConsulNamer` and farther: same
4. `mesh.InterpreterService.toBoundTreeRspEv`: transform to `VarEventStream.End(Throw(e))` (terminal event, containing either value or error)

  wire

5. `mesh.Client.streamActivity`: transforms End event into `Activity.Failed[NameTree[Name]]`
6. `Dst`: transforms to `Activity.Failed[BoundTree]`
7. `DynBoundFactory`: internal state transitions from Pending -> Failed. On `apply()` factory will return `Future.exception` (DunBoundFactory:81)
8. `ServiceFactoryCache`: wraps `DynBoundFactory` decorated with `NoBrokersAvailableException` handling into `IdlingFactory`
9. `IdlingFactory`: returns same `Future.exception`. Value remains in cache for TTL if idle
10. `RoutingService`: invokes `ServiceFactoryCache` and returns `Future.exception` as response. Every invocation resets idle timer, so as long as there are requests, `ServiceFactory` remains "stuck" in cache.

So it seems that
1. Even implementation of consul namer in master can put linkerd in "stuck" state (no previous state in namer, no index in response from consul). Obviously implementation from this PR will do this on any error. To get out of the "stuck" state linkerd has to be restarted or failing activity should be evicted from cache.
2. Considering that activities are memoized in namer, we also put namerd in "stuck" state. Looks like we should never interrupt a loop in activity (except for cases of explicit interruption). Theoretically[*], it can recover on its own, if all observes to activity are gone and activity is closed. Then the next observer will start the loop again.

Please, let me know if I missed something.

[*] during my experiments, I noticed that activity is not always closed as soon as all supposed observers are gone. I'd guess, there's another cache somewhere (maybe in http interface?). I had no time to dig into this however.

I'm going to change code to retry with back-off in case of exception.
I believe I should also change the case, when consul returns no index, but let me know if I misunderstood something. (Btw, why such case even exists? I understand, we can't do long poll on consul w/o index, but can't we just continue polling with some constant delay?)

adleong commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -145,6 +143,12 @@ private[consul] object SvcAddr {
     }
   }
 
+  private[this] def sleepAndDo[A](backoffs: Stream[Duration])(action: Function1[Stream[Duration], Future[A]]): Future[A] = {
```
I don't think we need this method.

adleong commented on namer/consul/src/test/scala/io/buoyant/namer/consul/ConsulNamerTest.scala:
```diff
@@ -106,17 +107,19 @@ class ConsulNamerTest extends FunSuite with Awaits {
   }
 
-  test("Namer goes pending when dc does not exist") {
+  test("Namer stops watching address when dc does not exist and there's no previous state") {
```
Given my comment above, I'm not sure this is the desirable behavior...

adleong commented on namer/consul/src/test/scala/io/buoyant/namer/consul/SvcAddrTest.scala:
```diff
@@ -0,0 +1,90 @@
+package io.buoyant.namer.consul
+
+import java.util.concurrent.atomic.AtomicInteger
+
```
group imports together

adleong commented on namer/consul/src/test/scala/io/buoyant/namer/consul/SvcAddrTest.scala:
```diff
@@ -0,0 +1,90 @@
+
+    // when
+    val addr: Var[Addr] = SvcAddr(api, backoffs, "dc1", SvcKey("svc", None), None, None, None, Map.empty, Stats(NullStatsReceiver))
+    addr.changes.respond { s => println(s) }
```
remove printlns please

adleong commented on namer/consul/src/main/scala/io/buoyant/namer/consul/LookupCache.scala:
```diff
@@ -1,13 +1,18 @@
 import com.twitter.util._
 import io.buoyant.consul.v1
-import io.buoyant.namer.Metadata
+
```
Linkerd style is to group all imports together (no blank lines) in alphabetical order

adleong commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -21,6 +22,8 @@ private[consul] object SvcAddr {
   private[this] val ServiceRelease =
     new Exception("service observation released") with NoStackTrace
 
+  implicit private[this] val timer: Timer = DefaultTimer
```
This will be more testable if you pass the Timer in as a param into apply.  This allows us to pass in a MockTimer in tests.

adleong commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -59,9 +63,9 @@ private[consul] object SvcAddr {
       stats.opens.incr()
       // last good state - if an error occurs, fall back to this if
```
I think this comment needs to be updated.  As far as I can tell, `lastGood` is only used for logging purposes now

hawkw commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -76,40 +77,19 @@ private[consul] object SvcAddr {
+              failureLogLevel,
+              "consul datacenter '%s' service '%s' observation error %s." +
+                " Last known state is %s",
+              datacenter, key.name, e, lastGood
```
Is the `lastGood` variable still necessary, or can we just use the current state of the activity for logging the last good state?

edio replied:
In general I agree, that storing same state in 2 places isn't very pretty.

However I don't see how can we get rid of `lastGood` if we want to log last state. We handle errors inside the loop and there we only have access to `Updatable`. Even conceptually it's tricky: closure with this loop is an argument to `Var`'s "constructor", so to get current value of `Var` one would need to populate a reference to yet unconstructed `Var` to the closure that is used to construct the `Var` instance.

We still can omit the state while logging. Imo it may be usefull during troubleshooting if any sort of outage happens, but it's something super important.

hawkw replied:
I definitely agree that logging the state is likely to be useful, and I think you're right that there's no easy way to avoid having the `lastGood` variable --- just wanted to make sure there wasn't an obvious way to remove it.

hawkw commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -76,40 +77,19 @@ private[consul] object SvcAddr {
+            )
+            val backoff #:: nextBackoffs = backoffs
+            // subsequent errors are logged as DEBUG
+            Future.sleep(backoff).before(loop(None, nextBackoffs, Level.DEBUG))
```
Passing the log level so that it's changed to debug on subsequent calls is very elegant, I like that 👍

Ashald commented:
I'm trying to get to review this PR but cannot make it out of meetings.. Hope to be done today.

Ashald commented:
In `LookupCache` we still have `agentConfig` that relies on `ConsulApi`'s retries. I wonder if it makes sense to generalize back-offs approach used in the namer and use it for that call as well. I think I saw someone suggesting getting rid of "InfiniteRetryFilter" for `ConsulApi` and if there were to happen we would need to reconsider how we handle `agentConfig`.

edio commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -50,18 +52,17 @@ private[consul] object SvcAddr {
-      // a good state was seen.
-      @volatile var lastGood: Option[Addr] = None
+      // log last observed state of the service to help in troubleshooting
+      @volatile var lastGood: Addr = Addr.Pending
```
Anyway, I read further and I agree that we should avoid storing state in volatile var, especially considering that it's only for logging. Will push update shortly

edio commented on namer/consul/src/main/scala/io/buoyant/namer/consul/SvcAddr.scala:
```diff
@@ -76,40 +77,19 @@ private[consul] object SvcAddr {
             )
```
@Ashald  we currently count closes (see Closable).
Ideally, this should correspond  to number of entrances into this branch (however, there's #1820 so it's not that simple). Having counter still may be a good idea, I'm not sure, whether we need to add this counter as part of this PR.

edio commented:
@Ashald I addressed your comments in ammended commit.
And since we started working on code readability, I also took the liberty to change `state() = ...` to `state.update(...)` as the latter is much more readable imo. I hope that's ok.
